# Sprocket Central Pty Ltd Customer Analysis Project

## Goals:
The primary objective of this project was to scrutinize datasets provided by Sprocket Central Pty Ltd, identify any data quality issues, and strategize on targeting the right customers. The project was segmented into distinct phases: data quality assessment, outlining a customer targeting approach, and developing a dashboard for presenting the findings.

## Project Description:

**Data Quality Assessment:** Comprehensive identification of data quality issues, such as missing values, inconsistencies, and potential outliers.  
Outcome: A clear communication was drafted and sent to the client, highlighting all the identified data quality concerns, providing them with a transparent view of the current state of their data.

**Presentation "Customer Targeting Strategy":** Creation of a PowerPoint presentation detailing the approach to be undertaken. The strategy was split into three core phases: Data Exploration, Model Development, and Interpretation.  
Outcome: A structured approach to sift through the dataset, develop predictive models, and extract meaningful insights about potential high-value customers.

**Dashboard Development:** A comprehensive dashboard was crafted, summarizing the data analysis results. This dashboard was tailored to present who, among the new 1000 customers, Sprocket Central Pty Ltd should specifically target.  
Outcome: An interactive, user-friendly dashboard that provides a clear visual representation of the key customers to target, backed by data-driven insights.

Links:  
1. [Transactions](#Transactions)
2. [NewCustomerList](#NewCustomerList)
3. [CustomerDemographic](#CustomerDemographic)
4. [CustomerAddress](#CustomerAddress)

In [ ]:
import pandas as pd

In [63]:
df_transactions = pd.read_excel('KPMG1.xlsx', sheet_name = 'Transactions')
df_new_customers = pd.read_excel('KPMG1.xlsx', sheet_name = 'NewCustomerList')
df_custom_demogr = pd.read_excel('KPMG1.xlsx', sheet_name = 'CustomerDemographic')
df_custom_address = pd.read_excel('KPMG1.xlsx', sheet_name = 'CustomerAddress')

/var/folders/th/zkxn724n2s12m0g5bscf4trh0000gp/T/ipykernel_45714/1351205394.py:2: FutureWarning: Inferring datetime64[ns] from data containing strings is deprecated and will be removed in a future version. To retain the old behavior explicitly pass Series(data, dtype=datetime64[ns])
  df_new_customers = pd.read_excel('KPMG1.xlsx', sheet_name = 'NewCustomerList')
/var/folders/th/zkxn724n2s12m0g5bscf4trh0000gp/T/ipykernel_45714/1351205394.py:3: FutureWarning: Inferring datetime64[ns] from data containing strings is deprecated and will be removed in a future version. To retain the old behavior explicitly pass Series(data, dtype=datetime64[ns])
  df_custom_demogr = pd.read_excel('KPMG1.xlsx', sheet_name = 'CustomerDemographic')


# Transactions

## Accuracy

In [20]:
df_transactions.head()

transaction_id  product_id  customer_id transaction_date  online_order  \
0           12826          11         3016       2017-12-30           1.0   
1           17394          20          933       2017-12-30           0.0   
2            9158          88           86       2017-12-30           1.0   
3           16258          14         2055       2017-12-30           1.0   
4            7311           1         1328       2017-12-30           0.0   

  order_status           brand product_line product_class product_size  \
0     Approved   Trek Bicycles     Standard        medium        small   
1     Approved   Trek Bicycles     Standard        medium        small   
2     Approved  Norco Bicycles     Standard          high        small   
3     Approved           Solex     Standard          high        large   
4     Approved  Giant Bicycles     Standard        medium       medium   

   list_price  standard_cost product_first_sold_date  
0     1775.81        1580.47              2014-03-03  
1     1775.81        1580.47              1996-04-05  
2     1661.92        1479.11              1993-07-15  
3     1842.92        1105.75              1995-10-24  
4     1403.50         954.82              2012-09-15

In [22]:
df_transactions.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20000 entries, 0 to 19999
Data columns (total 13 columns):
 #   Column                   Non-Null Count  Dtype         
---  ------                   --------------  -----         
 0   transaction_id           20000 non-null  int64         
 1   product_id               20000 non-null  int64         
 2   customer_id              20000 non-null  int64         
 3   transaction_date         20000 non-null  datetime64[ns]
 4   online_order             19640 non-null  float64       
 5   order_status             20000 non-null  object        
 6   brand                    19803 non-null  object        
 7   product_line             19803 non-null  object        
 8   product_class            19803 non-null  object        
 9   product_size             19803 non-null  object        
 10  list_price               20000 non-null  float64       
 11  standard_cost            19803 non-null  float64       
 12  product_first_sold_date  19803 n

In [41]:
fp_start = min(df_transactions.product_first_sold_date)
fp_end = max(df_transactions.product_first_sold_date)
print("First purchase dates from", fp_start, "to", fp_end)

tr_start = min(df_transactions.transaction_date)
tr_end = max(df_transactions.transaction_date)
print("Transaction dats from", tr_start, "to", tr_end)

First purchase dates from 1991-01-21 00:00:00 to 2016-12-06 00:00:00
Transaction dats from 2017-01-01 00:00:00 to 2017-12-30 00:00:00


In [102]:
df_transactions.describe()

transaction_id   product_id   customer_id  online_order    list_price  \
count    20000.000000  20000.00000  20000.000000  19640.000000  20000.000000   
mean     10000.500000     45.36465   1738.246050      0.500458   1107.829449   
std       5773.647028     30.75359   1011.951046      0.500013    582.825242   
min          1.000000      0.00000      1.000000      0.000000     12.010000   
25%       5000.750000     18.00000    857.750000      0.000000    575.270000   
50%      10000.500000     44.00000   1736.000000      1.000000   1163.890000   
75%      15000.250000     72.00000   2613.000000      1.000000   1635.300000   
max      20000.000000    100.00000   5034.000000      1.000000   2091.470000   

       standard_cost  
count   19803.000000  
mean      556.046951  
std       405.955660  
min         7.210000  
25%       215.140000  
50%       507.580000  
75%       795.100000  
max      1759.850000

In [103]:
print(df_transactions.customer_id.nunique())

3494


Everything looks accurate, except one id


**transaction_id:** values from 1 to 20,000  
**product_id:** values from 0 to 100, **id = 0 - strange**  
**customer_id:** values from 1 to 5034, but only 3494 unique   
**transaction_date:** dates from 2017-01-01 to 2017-12-30  
**online_order:** 1/online or 0/offline order  
**list_price:** prices from 12.01 to 2091.47  
**standard_cost:** costs from 7.21 to 1759.85  
**product_first_sold_date:** dates from 1991-01-21 to 2016-12-06  

## Completeness

In [77]:
missing_values_transactions = df_transactions.isnull().sum()

print(df_transactions.isna().sum()/len(df_transactions),
      missing_values_transactions)

transaction_id             0.00000
product_id                 0.00000
customer_id                0.00000
transaction_date           0.00000
online_order               0.01800
order_status               0.00000
brand                      0.00985
product_line               0.00985
product_class              0.00985
product_size               0.00985
list_price                 0.00000
standard_cost              0.00985
product_first_sold_date    0.00985
dtype: float64 transaction_id               0
product_id                   0
customer_id                  0
transaction_date             0
online_order               360
order_status                 0
brand                      197
product_line               197
product_class              197
product_size               197
list_price                   0
standard_cost              197
product_first_sold_date    197
dtype: int64


### Data has the following missing values:
**online_order:** 360, or 1,8%  
**brand:** 197, or ~0,001%  
**product_line:** 197, or ~0,001%  
**product_class:** 197, or ~0,001%  
**product_size:** 197, or ~0,001%  
**standard_cost:** 197, or ~0,001%  
**product_first_sold_date:** 197, or ~0,001%  

## Consistency

In [82]:
columns_transactions = ['order_status', 'brand', 'product_line', 'product_class', 'product_size']
unique_values_transactions = {}

for column in columns_transactions:
    unique_values_transactions[column] = df_transactions[column].value_counts()

unique_values_transactions

{'order_status': Approved     19821
 Cancelled      179
 Name: order_status, dtype: int64,
 'brand': Solex             4253
 Giant Bicycles    3312
 WeareA2B          3295
 OHM Cycles        3043
 Trek Bicycles     2990
 Norco Bicycles    2910
 Name: brand, dtype: int64,
 'product_line': Standard    14176
 Road         3970
 Touring      1234
 Mountain      423
 Name: product_line, dtype: int64,
 'product_class': medium    13826
 high       3013
 low        2964
 Name: product_class, dtype: int64,
 'product_size': medium    12990
 large      3976
 small      2837
 Name: product_size, dtype: int64}

It looks good. The large number of medium sizes and standard/universal models reflects the real-world distribution.  
**order_status:**  
- Approved: 19,821  
- Cancelled: 179  

**brand:**  
- Solex: 4,253  
- Giant Bicycles: 3,312  
- WeareA2B: 3,295  
- OHM Cycles: 3,043  
- Trek Bicycles: 2,990  
- Norco Bicycles: 2,910 

**product_line:**  
- Standard: 14,176  
- Road: 3,970  
- Touring: 1,234  
- Mountain: 423  

**product_class:**  
- Medium: 13,826  
- High: 3,013  
- Low: 2,964  

**product_size:**  
- Medium: 12,990  
- Large: 3,976  
- Small: 2,837


## Currency, Relevancy, Validity and Uniqueness

In [130]:
print("Dates from", tr_start, "to", tr_end)

Dates from 2017-01-01 00:00:00 to 2017-12-30 00:00:00


In [144]:
id0 = df_transactions[(df_transactions["product_id"] == 0) & 
                      (df_transactions["brand"].isnull()) & 
                      (df_transactions["product_line"].isnull()) &
                      (df_transactions["product_class"].isnull()) &
                      (df_transactions["standard_cost"].isnull()) &
                      (df_transactions["product_first_sold_date"].isnull())]
id0

transaction_id  product_id  customer_id transaction_date  online_order  \
155              5261           0         3322       2017-12-28           0.0   
156              5862           0         2404       2017-12-28           0.0   
263             19341           0          443       2017-12-26           1.0   
322             13011           0         1492       2017-12-25           1.0   
652              7235           0         2555       2017-12-19           1.0   
...               ...         ...          ...              ...           ...   
19497            6269           0         2121       2017-01-10           0.0   
19498           13369           0         1010       2017-01-10           1.0   
19719           10960           0          151       2017-01-06           1.0   
19841            4128           0         3242       2017-01-04           1.0   
19842           19206           0         2122       2017-01-04           1.0   

      order_status brand product_line product_class product_size  list_price  \
155       Approved   NaN          NaN           NaN          NaN     1563.94   
156       Approved   NaN          NaN           NaN          NaN     1672.07   
263       Approved   NaN          NaN           NaN          NaN      744.54   
322       Approved   NaN          NaN           NaN          NaN     1292.13   
652       Approved   NaN          NaN           NaN          NaN     1709.26   
...            ...   ...          ...           ...          ...         ...   
19497     Approved   NaN          NaN           NaN          NaN     1502.93   
19498     Approved   NaN          NaN           NaN          NaN     1315.16   
19719     Approved   NaN          NaN           NaN          NaN     1624.69   
19841     Approved   NaN          NaN           NaN          NaN      639.38   
19842     Approved   NaN          NaN           NaN          NaN     1615.62   

       standard_cost product_first_sold_date  
155              NaN                     NaT  
156              NaN                     NaT  
263              NaN                     NaT  
322              NaN                     NaT  
652              NaN                     NaT  
...              ...                     ...  
19497            NaN                     NaT  
19498            NaN                     NaT  
19719            NaN                     NaT  
19841            NaN                     NaT  
19842            NaN                     NaT  

[197 rows x 13 columns]

In [145]:
df_transactions.duplicated().sum()

0

**All rows with missing brand, product_line and etc. are with id = 0**

**The data covers one time period, most prices are positive, with only 1 repetitive anomaly. No duplicates were found.    
All provided data is relevant for analysis**

# NewCustomerList

## Accuracy

In [64]:
df_new_customers.head()

first_name   last_name  gender  past_3_years_bike_related_purchases  \
0    Marinna    Kauschke  Female                                   21   
1       Olia   O' Mullan  Female                                   77   
2   Brigitte    Whellams  Female                                   67   
3        Ivy        Farr  Female                                   56   
4   Beverlee  Ungerechts  Female                                   49   

         DOB                       job_title job_industry_category  \
0 1973-03-15                 Sales Associate    Financial Services   
1 1973-03-24               Account Executive                Health   
2 1973-05-09  Payment Adjustment Coordinator                   NaN   
3 1973-07-03             Office Assistant IV                    IT   
4 1973-10-03                  Civil Engineer         Manufacturing   

      wealth_segment deceased_indicator owns_car  ...  state    country  \
0  Affluent Customer                  N      Yes  ...    VIC  Australia   
1      Mass Customer                  N       No  ...    NSW  Australia   
2      Mass Customer                  N      Yes  ...    NSW  Australia   
3     High Net Worth                  N       No  ...    NSW  Australia   
4      Mass Customer                  N       No  ...    QLD  Australia   

   property_valuation Unnamed: 16 Unnamed: 17  Unnamed: 18  Unnamed: 19  \
0                   3        0.69      0.8625       0.8625      0.86250   
1                   7        0.41      0.4100       0.4100      0.34850   
2                   8        1.02      1.2750       1.2750      1.08375   
3                  11        0.86      0.8600       1.0750      1.07500   
4                   7        0.46      0.4600       0.4600      0.39100   

   Unnamed: 20  Rank     Value  
0        #REF!   595  0.775625  
1        #REF!   904  0.500000  
2        #REF!   188  1.156250  
3        #REF!   530  0.828750  
4        #REF!   617  0.754375  

[5 rows x 23 columns]

In [72]:
df_new_customers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 23 columns):
 #   Column                               Non-Null Count  Dtype         
---  ------                               --------------  -----         
 0   first_name                           1000 non-null   object        
 1   last_name                            971 non-null    object        
 2   gender                               1000 non-null   object        
 3   past_3_years_bike_related_purchases  1000 non-null   int64         
 4   DOB                                  983 non-null    datetime64[ns]
 5   job_title                            894 non-null    object        
 6   job_industry_category                835 non-null    object        
 7   wealth_segment                       1000 non-null   object        
 8   deceased_indicator                   1000 non-null   object        
 9   owns_car                             1000 non-null   object        
 10  tenure       

In [73]:
df_new_customers.describe()

past_3_years_bike_related_purchases       tenure     postcode  \
count                          1000.000000  1000.000000  1000.000000   
mean                             49.836000    11.388000  3019.227000   
std                              27.796686     5.037145   848.895767   
min                               0.000000     0.000000  2000.000000   
25%                              26.750000     7.000000  2209.000000   
50%                              51.000000    11.000000  2800.000000   
75%                              72.000000    15.000000  3845.500000   
max                              99.000000    22.000000  4879.000000   

       property_valuation  Unnamed: 16  Unnamed: 17  Unnamed: 18  Unnamed: 19  \
count         1000.000000  1000.000000  1000.000000  1000.000000  1000.000000   
mean             7.397000     0.741490     0.833278     0.936012     0.865490   
std              2.758804     0.204953     0.250907     0.296562     0.285798   
min              1.000000     0.400000     0.400000     0.400000     0.348500   
25%              6.000000     0.560000     0.625000     0.700000     0.637500   
50%              8.000000     0.740000     0.810000     0.912500     0.833000   
75%              9.000000     0.920000     1.025000     1.150000     1.062500   
max             12.000000     1.100000     1.375000     1.718750     1.718750   

              Rank        Value  
count  1000.000000  1000.000000  
mean    498.819000     0.881714  
std     288.810997     0.293525  
min       1.000000     0.340000  
25%     250.000000     0.649531  
50%     500.000000     0.860000  
75%     750.250000     1.075000  
max    1000.000000     1.718750

In [74]:
start_dob = min(df_new_customers.DOB)
end_dob = max(df_new_customers.DOB)
print("Dates of births from", start_dob, "to", end_dob)

Dates of births from 1938-06-08 00:00:00 to 2002-02-27 00:00:00


Everything looks accurate

**past_3_years_bike_related_purchase:** values from 0 to 99  
**tenure:** values (years) from 0 to 22  
**postcode:** values from 2000 to 4879  
**property_valuation:** values from 1 to 12  
**Rank:** values from 1 to 1000  
**Value:** values from 0.34 to 1.72  

## Completeness

In [80]:
missing_values_new_cust = df_new_customers.isnull().sum()

print(df_new_customers.isna().sum()/len(df_new_customers)*100,
      missing_values_new_cust)

first_name                              0.0
last_name                               2.9
gender                                  0.0
past_3_years_bike_related_purchases     0.0
DOB                                     1.7
job_title                              10.6
job_industry_category                  16.5
wealth_segment                          0.0
deceased_indicator                      0.0
owns_car                                0.0
tenure                                  0.0
address                                 0.0
postcode                                0.0
state                                   0.0
country                                 0.0
property_valuation                      0.0
Unnamed: 16                             0.0
Unnamed: 17                             0.0
Unnamed: 18                             0.0
Unnamed: 19                             0.0
Unnamed: 20                             0.0
Rank                                    0.0
Value                           

### Data has the following missing values:
**last_name:** 29, or 2.5%  
**DOB:** 17, or 1,7%  
**job_title:** 106, or 10,6%  
**job_industry_category:** 165, or 16,5%  

## Consistency

In [83]:
columns_new_customer_list = ['gender', 'job_title', 'job_industry_category', 'wealth_segment', 'deceased_indicator', 'owns_car', 'state']
unique_values_new_customer_list = {}

for column in columns_new_customer_list:
    unique_values_new_customer_list[column] = df_new_customers[column].value_counts()

unique_values_new_customer_list

{'gender': Female    513
 Male      470
 U          17
 Name: gender, dtype: int64,
 'job_title': Associate Professor            15
 Environmental Tech             14
 Software Consultant            14
 Chief Design Engineer          13
 Senior Sales Associate         12
                                ..
 Safety Technician II            1
 Computer Systems Analyst IV     1
 Database Administrator I        1
 Staff Accountant III            1
 Software Test Engineer III      1
 Name: job_title, Length: 184, dtype: int64,
 'job_industry_category': Financial Services    203
 Manufacturing         199
 Health                152
 Retail                 78
 Property               64
 IT                     51
 Entertainment          37
 Argiculture            26
 Telecommunications     25
 Name: job_industry_category, dtype: int64,
 'wealth_segment': Mass Customer        508
 High Net Worth       251
 Affluent Customer    241
 Name: wealth_segment, dtype: int64,
 'deceased_indicator': N    

In [85]:
print(df_new_customers.job_title.nunique())
print(df_new_customers.job_industry_category.nunique())

184
9



**gender:**  
- Female: 513  
- Male: 470
- **U: 17 – maybe other or NA?**

**job_title:**  
- 184 unique titles

**job_industry_category:**  
- 9 unique categories

As for **job_industry_category** some cells are n/a, while the in others columns of file cells without info are blank , it's better to choose one way

**wealth_segment:**  
- Mass Customer: 508  
- High Net Worth: 251  
- Affluent Customer: 241  

**deceased_indicator:**  
- All N, everybody is alive

**owns_car:**
- No: 507  
- Yes: 493  

**state:**
- NSW: 506
- VIC: 266
- QLD: 228


## Currency, Relevancy, Validity and Uniqueness

In [88]:
print("Dates of births from", start_dob, "to", end_dob)

Dates of births from 1938-06-08 00:00:00 to 2002-02-27 00:00:00


In [92]:
negative_values_new_customer_list = (df_new_customers[['past_3_years_bike_related_purchases', 'tenure', 'postcode', 'property_valuation']] < 0).sum()

# Check if there are any DOB dates in the future (beyond the current date) for "NewCustomerList"
future_dob_dates_new_customer_list = (df_new_customers['DOB'] > pd.Timestamp.now()).sum()

negative_values_new_customer_list, future_dob_dates_new_customer_list


(past_3_years_bike_related_purchases    0
 tenure                                 0
 postcode                               0
 property_valuation                     0
 dtype: int64,
 0)

In [93]:
df_new_customers.duplicated().sum()

0

**All values are positive and dates look real, no anomalies or duplicates.   
All provided data is relevant for analysis**

# CustomerDemographic

## Accuracy

In [94]:
df_custom_demogr.head()

customer_id      first_name  last_name  gender  \
0            1         Laraine  Medendorp       F   
1            2             Eli    Bockman    Male   
2            3           Arlin     Dearle    Male   
3            4          Talbot        NaN    Male   
4            5  Sheila-kathryn     Calton  Female   

   past_3_years_bike_related_purchases        DOB               job_title  \
0                                   93 1953-10-12     Executive Secretary   
1                                   81 1980-12-16  Administrative Officer   
2                                   61 1954-01-20      Recruiting Manager   
3                                   33 1961-10-03                     NaN   
4                                   56 1977-05-13           Senior Editor   

  job_industry_category     wealth_segment deceased_indicator  \
0                Health      Mass Customer                  N   
1    Financial Services      Mass Customer                  N   
2              Property      Mass Customer                  N   
3                    IT      Mass Customer                  N   
4                   NaN  Affluent Customer                  N   

                                             default owns_car  tenure  
0                                                 "'      Yes    11.0  
1                       <script>alert('hi')</script>      Yes    16.0  
2                                2018-02-01 00:00:00      Yes    15.0  
3  () { _; } >_[$($())] { touch /tmp/blns.shellsh...       No     7.0  
4                                                NIL      Yes     8.0

In [95]:
df_custom_demogr.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4000 entries, 0 to 3999
Data columns (total 13 columns):
 #   Column                               Non-Null Count  Dtype         
---  ------                               --------------  -----         
 0   customer_id                          4000 non-null   int64         
 1   first_name                           4000 non-null   object        
 2   last_name                            3875 non-null   object        
 3   gender                               4000 non-null   object        
 4   past_3_years_bike_related_purchases  4000 non-null   int64         
 5   DOB                                  3913 non-null   datetime64[ns]
 6   job_title                            3494 non-null   object        
 7   job_industry_category                3344 non-null   object        
 8   wealth_segment                       4000 non-null   object        
 9   deceased_indicator                   4000 non-null   object        
 10  default     

In [96]:
df_custom_demogr.describe()

customer_id  past_3_years_bike_related_purchases       tenure
count  4000.000000                          4000.000000  3913.000000
mean   2000.500000                            48.890000    10.657041
std    1154.844867                            28.715005     5.660146
min       1.000000                             0.000000     1.000000
25%    1000.750000                            24.000000     6.000000
50%    2000.500000                            48.000000    11.000000
75%    3000.250000                            73.000000    15.000000
max    4000.000000                            99.000000    22.000000

Everything looks accurate

**customer_id:**: Values from 1 to 4000, while in Transactions it's from 1 to 5034  
**past_3_years_bike_related_purchases:** values from 0 to 99  
**tenure:** values (years) from 1 to 22


## Completeness

In [98]:
missing_values_customer_demogr = df_custom_demogr.isnull().sum()


print(df_custom_demogr.isna().sum()/len(df_custom_demogr)*100,
      missing_values_customer_demogr)

customer_id                             0.000
first_name                              0.000
last_name                               3.125
gender                                  0.000
past_3_years_bike_related_purchases     0.000
DOB                                     2.175
job_title                              12.650
job_industry_category                  16.400
wealth_segment                          0.000
deceased_indicator                      0.000
default                                 7.550
owns_car                                0.000
tenure                                  2.175
dtype: float64 customer_id                              0
first_name                               0
last_name                              125
gender                                   0
past_3_years_bike_related_purchases      0
DOB                                     87
job_title                              506
job_industry_category                  656
wealth_segment                           0


### Data has the following missing values:
**last_name:** 125, or ~3.1%  
**DOB:** 87, or ~2.2%  
**job_title:** 506, or ~12.7%  
**job_industry_category:** 656, or 16,4%  
**default:** 302, or ~7,6%  
**tenure:** 87, or ~2,2%  

## Consistency

In [100]:
columns_customer_demogr = ['gender', 'job_title', 'job_industry_category', 'wealth_segment', 'deceased_indicator', 'owns_car']
unique_values_customer_demogr = {}

for column in columns_customer_demogr:
    unique_values_customer_demogr[column] = df_custom_demogr[column].value_counts()

unique_values_customer_demogr

{'gender': Female    2037
 Male      1872
 U           88
 F            1
 Femal        1
 M            1
 Name: gender, dtype: int64,
 'job_title': Business Systems Development Analyst    45
 Tax Accountant                          44
 Social Worker                           44
 Internal Auditor                        42
 Recruiting Manager                      41
                                         ..
 Database Administrator I                 4
 Health Coach I                           3
 Health Coach III                         3
 Research Assistant III                   3
 Developer I                              1
 Name: job_title, Length: 195, dtype: int64,
 'job_industry_category': Manufacturing         799
 Financial Services    774
 Health                602
 Retail                358
 Property              267
 IT                    223
 Entertainment         136
 Argiculture           113
 Telecommunications     72
 Name: job_industry_category, dtype: int64,
 'wealth_se

In [101]:
print(df_custom_demogr.job_industry_category.nunique())

9


**gender:**
- Female: 2037  
- Male: 1872  
- **U: 88 - again U**  
- **F: 1 - the only case with "F" for "Female"**  
- **Femal: 1 "Female", spelling**   
- **M: 1 the only case with "M" for "Male"**  

**job_title:**
- 195 unique titles  

**job_industry_category:**
- 9 job category  

**wealth_segment:**
- Mass Customer: 2000
- High Net Worth: 1021
- Affluent Customer: 979

**deceased_indicator:**
- N: 3998 
- Y: 2 - only 2 dead clients

**owns_car:**
- Yes: 2024
- No: 1976

## Currency, Relevancy, Validity and Uniqueness

In [105]:
start_dob_demogr = df_custom_demogr.DOB.min()
end_dob_demogr = df_custom_demogr.DOB.max()

print("Dates of births from", start_dob_demogr, "to", end_dob_demogr)

Dates of births from 1843-12-21 00:00:00 to 2002-03-11 00:00:00


In [111]:
df_custom_demogr.DOB.dropna().sort_values()

33     1843-12-21
719    1931-10-23
1091   1935-08-22
3409   1940-09-22
2412   1943-08-11
          ...    
421    2002-01-06
2857   2002-01-09
3434   2002-01-15
1887   2002-01-26
65     2002-03-11
Name: DOB, Length: 3913, dtype: datetime64[ns]

In [115]:
negative_values_customer_demogr = (df_custom_demogr[['past_3_years_bike_related_purchases', 'tenure']] < 0).sum()
negative_values_customer_demogr

past_3_years_bike_related_purchases    0
tenure                                 0
dtype: int64

In [117]:
df_custom_demogr.duplicated().sum()

0

**All values are positive, no duplicates. One Date of Birth is outlier.  
"Default" column has no analytical value, moveover it contains part of malicious code. Other provided data is relevant for analysis.**

# CustomerAddress

## Accuracy

In [118]:
df_custom_address.head()

customer_id              address  postcode            state    country  \
0            1   060 Morning Avenue      2016  New South Wales  Australia   
1            2  6 Meadow Vale Court      2153  New South Wales  Australia   
2            4   0 Holy Cross Court      4211              QLD  Australia   
3            5  17979 Del Mar Point      2448  New South Wales  Australia   
4            6     9 Oakridge Court      3216              VIC  Australia   

   property_valuation  
0                  10  
1                  10  
2                   9  
3                   4  
4                   9

In [119]:
df_custom_address.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3999 entries, 0 to 3998
Data columns (total 6 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   customer_id         3999 non-null   int64 
 1   address             3999 non-null   object
 2   postcode            3999 non-null   int64 
 3   state               3999 non-null   object
 4   country             3999 non-null   object
 5   property_valuation  3999 non-null   int64 
dtypes: int64(3), object(3)
memory usage: 187.6+ KB


In [121]:
df_custom_address.describe()

customer_id     postcode  property_valuation
count  3999.000000  3999.000000         3999.000000
mean   2003.987997  2985.755939            7.514379
std    1154.576912   844.878364            2.824663
min       1.000000  2000.000000            1.000000
25%    1004.500000  2200.000000            6.000000
50%    2004.000000  2768.000000            8.000000
75%    3003.500000  3750.000000           10.000000
max    4003.000000  4883.000000           12.000000

Everything looks accurate

**customer_id:** values from 1 to 4003  
**postcode:** values from 2000 to 4883  
**property_valuation:** in the range from 1 to 12  

## Completeness

In [125]:
missing_values_custom_address = df_custom_address.isnull().sum()

missing_values_custom_address

customer_id           0
address               0
postcode              0
state                 0
country               0
property_valuation    0
dtype: int64

### Data has no missing values

## Consistency

In [127]:
columns_custom_address = ['address', 'state', 'country']
unique_values_custom_address = {}

for column in columns_custom_address:
    unique_values_custom_address[column] = df_custom_address[column].value_counts()

unique_values_custom_address['state'], unique_values_custom_address['country']


(NSW                2054
 VIC                 939
 QLD                 838
 New South Wales      86
 Victoria             82
 Name: state, dtype: int64,
 Australia    3999
 Name: country, dtype: int64)

**state:**
- NSW: 2054
- VIC: 939
- QLD: 838
- **New South Wales: 86 (Duplicate for "NSW")**
- **Victoria: 82 (Duplicate for "VIC")**  

**country:** Only Australia

## Currency, Relevancy, Validity and Uniqueness

**All values are positive and dates look real, no anomalies or duplicates. The only issue is with short state names.**   
**Also there are some number 0 streets, not sure if it's mistake.**  

**All provided data is relevant for analysis.**